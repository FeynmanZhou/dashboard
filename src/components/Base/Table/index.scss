@import '~scss/variables';

.table {
  &-body {
    transition: opacity $transition-speed;
  }

  table {
    width: 100%;
    border-collapse: collapse;
    border-spacing: 0;
    text-align: left;
    box-sizing: border-box;
    overflow: hidden;
  }
}

:global {
  .pi-table {
    &-thead > tr > th {
      position: relative;
      padding: 12px 24px;
      font-weight: 500;
      color: $title-color;
      background: $th-background-color;

      &::before {
        content: "";
        position: absolute;
        left: -1px;
        top: 8px;
        bottom: 8px;
        width: 0;
        border-right: 1px solid $border-light;
      }
    }

    &-tbody > tr{
      transition: all $transition-speed;

      &:hover {
        background-color: #fafafa;
        
        td {
          &:first-child {
            border-left-color: #fafafa;
          }
          &:last-child {
            border-right-color: #fafafa;
          }
        }
      }
    }

    &-tbody > tr > td {
      padding: 12px 0 12px 24px;
      line-height: 16px;
      border-bottom: 1px solid $border-lighter;
      transition: all $transition-speed;
      word-wrap:break-word;

      &:first-child {
        padding-left: 0;
        border-left: 48px solid #fff;
      }

      &:last-child {
        border-right: 48px solid #fff;
      }
    }

    &-placeholder {
      padding: 12px 0;
      text-align: center;
    }
  }
}

.selectionCheck {
  padding-left: 14px;
}
