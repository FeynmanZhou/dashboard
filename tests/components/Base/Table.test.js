import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Table from 'components/Base/Table';

configure({ adapter: new Adapter() });

const setup = (props = {}, type) => {
  const wrapper = type === 'mount' ? mount(<Table {...props}/>) : shallow(<Table {...props}/>);

  return {
    wrapper,
    props: wrapper.instance().props,
  };
};

describe('Base/Table', () => {
  const columns = [
    {
      title: 'Name', dataIndex: 'name', key: 'name', width: '50%',
    },
    {
      title: 'Age', dataIndex: 'age', key: 'age', width: '50%',
    },
  ];
  const dataSource = [
    {
      name: 'zhangsan', age: 20, key: '1',
    },
    {
      name: 'lisi', age: 22, key: '2',
    },    
  ];
  const rowSelection = {
    selectedRowKeys: [],
    onChange: jest.fn(),
    onSelect: jest.fn(),
    onSelectAll: jest.fn(),
  };  

  it('render without crash', () => {
    const { wrapper } = setup();
  });

  it('custom columns & data', () => {
    const { wrapper } = setup({ columns, dataSource }, 'mount');
    
    expect(wrapper.find('thead th').length).toEqual(columns.length);
    expect(wrapper.find('tbody tr').length).toEqual(dataSource.length);
  });

  it('call change', () => {
    const { wrapper } = setup({ columns, dataSource, rowSelection }, 'mount');
    const checkbox = wrapper.find('tbody tr').first().find('td').at(0).find('input[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });

    const mockChange = rowSelection.onChange;
    expect(wrapper.state().selectionDirty).toBeTruthy();
    expect(mockChange).toHaveBeenCalled();
    expect(mockChange.mock.calls[0][0]).toEqual(['1']);
  });

  it('call select one', () => {
    const { wrapper } = setup({ columns, dataSource, rowSelection }, 'mount');
    const checkbox = wrapper.find('tbody tr').first().find('td').at(0).find('input[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });

    const mockSelect = rowSelection.onSelect;
    expect(wrapper.state().selectionDirty).toBeTruthy();
    expect(mockSelect).toHaveBeenCalled();
    expect(mockSelect.mock.calls[0][1]).toEqual(dataSource[0]);
  });

  it('call select all', () => {
    const { wrapper } = setup({ columns, dataSource, rowSelection }, 'mount');
    const checkbox = wrapper.find('thead th').at(0).find('input[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });

    const mockSelectAll = rowSelection.onSelectAll;
    expect(wrapper.state().selectionDirty).toBeTruthy();
    expect(mockSelectAll).toHaveBeenCalled();
    expect(mockSelectAll.mock.calls[0][1]).toEqual(dataSource);
  });
});
