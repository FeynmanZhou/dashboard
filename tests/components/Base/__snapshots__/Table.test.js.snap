// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Table basic render 1`] = `
<div
  class="table"
>
  <div
    class="pi-table pi-table-scroll-position-left"
  >
    <div
      class="pi-table-content"
    >
      <div
        class="pi-table-body"
      >
        <table
          class=""
        >
          <colgroup>
            <col
              style="width:50%;min-width:50%"
            />
            <col
              style="width:50%;min-width:50%"
            />
          </colgroup>
          <thead
            class="pi-table-thead"
          >
            <tr>
              <th
                class=""
              >
                Name
              </th>
              <th
                class=""
              >
                Age
              </th>
            </tr>
          </thead>
          <tbody
            class="pi-table-tbody"
          >
            <tr
              class="pi-table-row  pi-table-row-level-0"
            >
              <td
                class=""
              >
                <span
                  class="pi-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                zhangsan
              </td>
              <td
                class=""
              >
                20
              </td>
            </tr>
            <tr
              class="pi-table-row  pi-table-row-level-0"
            >
              <td
                class=""
              >
                <span
                  class="pi-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                lisi
              </td>
              <td
                class=""
              >
                22
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
